### PR TITLE
Djhoward12 naaps 10 app rackgroup assignment

### PR DESCRIPTION
**Pull request by djhoward12**
Added the ability to assign RackGroups and Status tiles to span multiple spaces and place a Rack tile on top of a group of tiles. 

Closes: #10

### CODE REVIEW

File: nautobot_floor_plan/choices.py [hunk 1]
```diff
@@ -29,3 +29,15 @@ class AxisLabelsChoices(ChoiceSet):
...
+
+class AllocationTypeChoices(ChoiceSet):
+    """Choices for tile allocation type."""
+
+    RACK = "RACK"
+    RACKGROUP = "RACKGROUP"
+
+    CHOICES = (
+        (RACK, "RACK"),
+        (RACKGROUP, "RACKGROUP"),
```
Comment (glennmatthews) [inline nautobot_floor_plan/choices.py:42]: Not critical, but by convention this would normally be more like:

```suggestion
    RACK = "rack"
    RACKGROUP = "rackgroup"

    CHOICES = (
        (RACK, "Rack"),
        (RACKGROUP, "Rack Group"),
```

File: nautobot_floor_plan/migrations/0005_auto_20240621_1852.py
Comment (glennmatthews) [inline nautobot_floor_plan/migrations/0005_auto_20240621_1852.py:1]: Mind renaming this file to something more descriptive like `0005_add_rackgroup.py`  or similar?

File: nautobot_floor_plan/models.py [hunk 2]
```diff
@@ -145,8 +184,36 @@ def clean(self):
...
+                        {f"Rack {self.rack} must not extend beyond the boundary of the defined group tiles"}
+                    )
+                if y_max > oy_max or y_min < oy_min:
+                    raise ValidationError(
+                        {f"Rack {self.rack} must not extend beyond the boundary of the defined group tiles"}
+                    )
+                self.on_group_tile = True
+                if orack_group is not None:
+                    if orack_group != rack_group:
+                        # If rack is not in the correct rack group or rack with with a rack group is being placed on a status group tile?
```
Comment (glennmatthews) [inline nautobot_floor_plan/models.py:204]: I find this comment hard to follow, perhaps reword it?

File: nautobot_floor_plan/models.py [hunk 3]
```diff
@@ -145,8 +184,36 @@ def clean(self):
...
+                        raise ValidationError({"rack_group": f"Rack {self.rack} must belong to {orack_group}"})
```
Comment (glennmatthews) [inline nautobot_floor_plan/models.py:205]: Would this be as correct and any clearer?

```suggestion
                        raise ValidationError({"rack_group": f"Rack {self.rack} does not belong to {orack_group}"})
```

File: nautobot_floor_plan/models.py [hunk 4]
```diff
@@ -145,8 +184,36 @@ def clean(self):
...
+            if rack is None:
+                if x_max > ox_max or x_min < ox_min:
+                    # Allow for the extending of exising underlying Rack Group and Status tiles
+                    if orack_group is not None and allocation_type != AllocationTypeChoices.RACKGROUP:
+                        raise ValidationError({"Tile must not extend beyond the boundary of the defined group tiles"})
+                if y_max > oy_max or y_min < oy_min:
+                    # Allow for the extending of exising underlying Rack Group and Status tiles
+                    if orack_group is not None and allocation_type != AllocationTypeChoices.RACKGROUP:
+                        raise ValidationError({"Tile must not extend beyond the boundary of the defined group tiles"})
```
Comment (glennmatthews) [inline nautobot_floor_plan/models.py:214]: Not sure I understand the validation case that's being checked for here. Can you elaborate in the comment?

File: nautobot_floor_plan/svg.py [hunk 5]
```diff
@@ -174,117 +180,180 @@ def _draw_grid(self, drawing):
...
+
     def _draw_tile(self, drawing, tile):
         """Render an individual FloorPlanTile to the drawing."""
-        # Draw the square defining the bounds of this tile
-        self._draw_defined_tile(drawing, tile)
+        # functions to handle rack_group tiles and status tiles.
+        # Add text to status group and rack group tiles
+        self._draw_defined_rackgroup_tile(drawing, tile)
+        # Draw status grids with edit and delete buttons
+        self._draw_defined_status_tiles(drawing, tile)
```
Comment (glennmatthews) [inline nautobot_floor_plan/svg.py:254]: Why `rackgroup_tile` but `status_tiles` (singular vs plural)?

File: nautobot_floor_plan/svg.py [hunk 6]
```diff
@@ -174,117 +180,180 @@ def _draw_grid(self, drawing):
                     )
                 )
 
+    # draw edit and delete functionality to Status and Rackgroup tiles
+    def _draw_defined_status_tiles(self, drawing, tile):
+        """Render a tile based on its Status."""
```
Comment (glennmatthews) [inline nautobot_floor_plan/svg.py:185]: Could you update the docstring to clarify what this method is doing and how it differs from `_draw_status_tiles`?

File: nautobot_floor_plan/svg.py [hunk 7]
```diff
@@ -388,6 +479,62 @@ def _draw_rack_tile(self, drawing, tile):
...
+        link.add(
+            drawing.text(
+                "✎",
+                insert=(
+                    origin[0] + self.TILE_INSET + self.TEXT_LINE_HEIGHT / 2 + self.RACK_BUTON_OFFSET,
+                    origin[1] + self.TILE_INSET + self.TEXT_LINE_HEIGHT / 2 + self.GRID_OFFSET,
+                ),
+                class_="button-text",
+            )
+        )
```
Comment (glennmatthews) [inline nautobot_floor_plan/svg.py:504]: Would it be possible/appropriate to create some reusable helper functions for the "edit" and "delete" button rendering? Feels like a lot of repeated code for these.

File: nautobot_floor_plan/svg.py [hunk 8]
```diff
@@ -24,9 +24,15 @@ class FloorPlanSVG:
     CORNER_RADIUS = 6
     TILE_INSET = 2
     TEXT_LINE_HEIGHT = 16
-    GRID_OFFSET = 30
+    GRID_OFFSET = 26
     RACK_INSETS = (3 * TILE_INSET, 3 * TILE_INSET + TEXT_LINE_HEIGHT)
+    RACK_PADDING = 4
+    RACK_TILE_INSET = 3
     RACK_FRONT_DEPTH = 15
+    RACK_BUTON_OFFSET = 5
```
Comment (glennmatthews) [inline nautobot_floor_plan/svg.py:32]: `RACK_BUTTON_OFFSET`

File: nautobot_floor_plan/svg.py [hunk 9]
```diff
@@ -396,10 +543,16 @@ def render(self):
             width=self.floor_plan.x_size * self.GRID_SIZE_X + self.GRID_OFFSET + self.BORDER_WIDTH * 2,
             depth=self.floor_plan.y_size * self.GRID_SIZE_Y + self.GRID_OFFSET + self.BORDER_WIDTH * 2,
         )
+        # Draw Rack Groups and status boxes before the grid is created
+        logger.debug("Rendering underlying rack_group and status tiles...")
+        for tile in self.floor_plan.tiles.all():
+            self._draw_status_tiles(drawing, tile)
```
Comment (glennmatthews) [inline nautobot_floor_plan/svg.py:549]: Would something like `_draw_underlay_tiles` be a clearer name?

File: nautobot_floor_plan/tests/test_api_views.py [hunk 10]
```diff
@@ -46,31 +46,59 @@ class FloorPlanTileAPIViewTest(APIViewTestCases.APIViewTestCase):
 
     model = models.FloorPlanTile
     brief_fields = ["display", "id", "url", "x_origin", "x_size", "y_origin", "y_size"]
-    choices_fields = ["rack_orientation"]
+    choices_fields = ["rack_orientation", "allocation_type"]
     # TODO choices_fields = ["rack_orientation", "status"]
```
Comment (glennmatthews) [inline nautobot_floor_plan/tests/test_api_views.py:50]: Can remove this line I think.

File: nautobot_floor_plan/tests/test_models.py [hunk 11]
```diff
@@ -134,6 +137,26 @@ def test_create_floor_plan_tile_invalid_duplicate_rack(self):
...
+        # How about a rack without the correct rack group?
+        non_rack_group_rack = Rack.objects.create(name="Rack 2", status=self.active_status, location=self.floors[2])
+        with self.assertRaises(ValidationError):
+            models.FloorPlanTile(
+                floor_plan=self.floor_plans[2],
+                x_origin=1,
+                y_origin=1,
+                status=self.active_status,
+                rack=non_rack_group_rack,
+            ).validated_save()
```
Comment (glennmatthews) [inline nautobot_floor_plan/tests/test_models.py:158]: Would be good to see test coverage for each of the new checks in `group_tile_bounds()`.

**Review by glennmatthews — COMMENTED** [review]
Good stuff!

File: nautobot_floor_plan/models.py (hunk 4 above)
Comment (djhoward12) [reply nautobot_floor_plan/models.py:214]: Changing comment to
# RACKGROUP tiles can grow and shrink but not overlap other RACKGROUP tiles

File: nautobot_floor_plan/models.py (hunk 2 above)
Comment (djhoward12) [reply nautobot_floor_plan/models.py:204]: Changing comment to
```python
# Is tile assigned to a rack_group? Racks must be assigned to the same rack_group
```

File: nautobot_floor_plan/tests/test_api_views.py (hunk 10 above)
Comment (djhoward12) [reply nautobot_floor_plan/tests/test_api_views.py:50]: Unable to remove 49. Received the following error when running unittests
```
 All field names of choice fields for a given model serializer need to be manually added to self.choices_fields. If this is already the case, perhaps the serializer is implemented incorrectly?
```

Comment (glennmatthews) [reply nautobot_floor_plan/tests/test_api_views.py:50]: Sorry, I was actually referring to the commented-out line 50.

File: nautobot_floor_plan/svg.py (hunk 7 above)
Comment (djhoward12) [reply nautobot_floor_plan/svg.py:504]: This is now a edit and delete function and the function _draw_defined_status_tiles has been removed.

**Comment by joewesch** [comment]
So when you have a Floor Tile "group" that is bigger than a rack within that group (left side of first image) it makes sense to have an edit for both the rack tile and the group tile. However, if the tile takes up the same spaces (right side of first image and second image) then can we hide one of the sets of Edit/Delete buttons since they seem to take you to the same form/page?

Screenshot 2024-06-26 at 3 28 55 PM
Screenshot 2024-06-26 at 3 31 16 PM

File: docs/admin/release_notes/version_2.2.md [hunk 12]
```diff
@@ -0,0 +1,6 @@
+
+## [v2.2.0 (2024-06-28)](https://github.com/nautobot/nautobot-app-floor-plan/releases/tag/v2.2.0)
+
+### Added
+
+- [#10](https://github.com/nautobot/nautobot-app-floor-plan/issues/10) - Added the ability to allow a rackgroup or status tile to span multiple floortiles. Additionally rack tiles can be place on top of a group of status or rackgroup tiles.
```
Comment (joewesch) [inline docs/admin/release_notes/version_2.2.md:6]: I think this file gets generated automatically during the release process from town crier with the proper date and such. Not sure what would happen if it is already created, so I would probably just err on the side of caution and remove it for now.

**Comment by djhoward12** [comment]
> So when you have a Floor Tile "group" that is bigger than a rack within that group (left side of first image) it makes sense to have an edit for both the rack tile and the group tile. However, if the tile takes up the same spaces (right side of first image and second image) then can we hide one of the sets of Edit/Delete buttons since they seem to take you to the same form/page?
> 
> Screenshot 2024-06-26 at 3 28 55 PM Screenshot 2024-06-26 at 3 31 16 PM

This has been resolved with the latest update.

File: nautobot_floor_plan/models.py [hunk 13]
```diff
@@ -145,8 +184,38 @@ def clean(self):
...
+                    )
+                if y_max > oy_max or y_min < oy_min:
+                    raise ValidationError(
+                        {f"Rack {self.rack} must not extend beyond the boundary of the defined group tiles"}
+                    )
+                self.on_group_tile = True
+                if orack_group is not None:
+                    if orack_group != rack_group:
+                        # Is tile assigned to a rack_group? Racks must be assigned to the same rack_group
+                        raise ValidationError({"rack_group": f"Rack {self.rack} does not belong to {orack_group}"})
```
Comment (joewesch) [inline nautobot_floor_plan/models.py:205]: I talked with DJ on a change on the side, but putting it here for transparency.

If you have a Rack Group tile and you try to assign a tile on top of it with just a Rack (not a Rack Group) it will show up with this error message - even if the Rack in question is indeed in the Rack Group. The current thought is above this we can look to see if they didn't provide the Rack Group in the form and we can just replace it with `rack.rack_group` to "infer" the Rack Group that should have been assigned so we don't throw this error unless the Rack chosen isn't in the proper Rack Group for the larger tile.

File: nautobot_floor_plan/forms.py [hunk 14]
```diff
@@ -83,7 +83,12 @@ class FloorPlanTileForm(NautobotModelForm):
 
     floor_plan = DynamicModelChoiceField(queryset=models.FloorPlan.objects.all())
     rack = DynamicModelChoiceField(
-        queryset=Rack.objects.all(), required=False, query_params={"nautobot_floor_plan_floor_plan": "$floor_plan"}
+        queryset=Rack.objects.all(),
+        required=False,
+        query_params={"nautobot_floor_plan_floor_plan": "$floor_plan", "rack_group": "$rack_group"},
+    )
+    rack_group = DynamicModelChoiceField(
+        queryset=RackGroup.objects.all(), required=False, query_params={"nautobot_floor_plan_floor_plan": "$floor_plan"}
```
Comment (joewesch) [inline nautobot_floor_plan/forms.py:91]:
```suggestion
    rack_group = DynamicModelChoiceField(
        queryset=RackGroup.objects.all(),
        required=False,
        query_params={"nautobot_floor_plan_floor_plan": "$floor_plan", "racks": "$rack"},
```

This filters the Rack Group dropdown to only allow groups that includes the Rack selected (if it has been selected)

File: docs/user/app_getting_started.md [hunk 15]
```diff
@@ -32,7 +32,9 @@ After clicking "Create", you will be presented with a new floor plan render:
...
 
-You can assign a Status to each tile, and optionally assign a Rack as well as specifying the orientation of the Rack relative to the Floor Plan. You can also specify the size of a Tile if you want it to cover multiple "spaces" in the Floor Plan - this can be useful to document larger-than-usual Racks, or to mark entire sections of the Floor Plan as "Reserved" or "Unavailable".
+You can assign a Status to each tile, and optionally assign a Rack or Rackgroup as well as specifying the orientation of the Rack relative to the Floor Plan. You can also specify the size of a Tile if you want it to cover multiple "spaces" in the Floor Plan - this can be useful to document larger-than-usual Racks, or to mark entire sections of the Floor Plan as "Reserved" or "Unavailable". 
+
+You can place racks within the Status or Rackgroup tiles that are covering multiple spaces. It placing a Rack onto a Rackgroup tile, the rack must be added to the appropriate Rackgroup. Rackgroup and Status tiles that cover multiple "spaces" can be increased or reduced in size as long as they don't overlap with over status or Rackgroup tiles.
```
Comment (joewesch) [inline docs/user/app_getting_started.md:37]:
```suggestion
You can place racks within the Status or Rackgroup tiles that are covering multiple spaces. When placing a Rack onto a Rackgroup tile, the rack must be added to the appropriate Rackgroup. Rackgroup and Status tiles that cover multiple "spaces" can be increased or reduced in size as long as they don't overlap with over status or Rackgroup tiles.
```

File: nautobot_floor_plan/models.py [hunk 16]
```diff
@@ -119,23 +119,62 @@ class FloorPlanTile(PrimaryModel):
...
+    allocation_type = models.CharField(
+        choices=AllocationTypeChoices,
+        max_length=10,
+        blank=True,
+    )
+
+    on_group_tile = models.BooleanField(
+        default=False,
+        blank=True,
+        null=True,
```
[the remark this answers is not in the document]
Comment (glennmatthews) [reply nautobot_floor_plan/models.py:145]: Also, does this field actually need to be nullable?

File: nautobot_floor_plan/models.py [hunk 17]
```diff
@@ -145,8 +184,41 @@ def clean(self):
...
+                        # Is tile assigned to a rack_group? Racks must be assigned to the same rack_group
+                        raise ValidationError({"rack_group": f"Rack {self.rack} must belong to {orack_group}"})
+            if rack is None:
+                # RACKGROUP tiles can grow and shrink but not overlap other RACKGROUP tiles or Racks that are not assigned to the correct rackgroup
+                if x_max > ox_max or x_min < ox_min:
+                    if oallocation_type == AllocationTypeChoices.RACK and orack_group != rack_group:
+                        raise ValidationError("Tile overlaps a Rack that is not in the correct Rackgroup")
+                if y_max > oy_max or y_min < oy_min:
+                    if oallocation_type == AllocationTypeChoices.RACK and orack_group != rack_group:
+                        raise ValidationError("Tile overlaps a Rack that is not in the correct Rackgroup")
```
Comment (glennmatthews) [inline nautobot_floor_plan/models.py:217]:
```suggestion
                        raise ValidationError("Tile overlaps a Rack that is not in the specified RackGroup")
                if y_max > oy_max or y_min < oy_min:
                    if oallocation_type == AllocationTypeChoices.RACK and orack_group != rack_group:
                        raise ValidationError("Tile overlaps a Rack that is not in the specified RackGroup")
```

File: nautobot_floor_plan/templates/nautobot_floor_plan/floorplantile_retrieve.html [hunk 18]
```diff
@@ -37,10 +37,18 @@
                     </span>
                 </td>
             </tr>
+            <tr>
+                <td>Tile Type</td>
+                <td>{{ object.allocation_type }}</td>
```
Comment (glennmatthews) [inline nautobot_floor_plan/templates/nautobot_floor_plan/floorplantile_retrieve.html:42]: Possibly:

```suggestion
                <td>{{ object.get_allocation_type_display|placeholder }}</td>
```

since this field is blankable.

File: nautobot_floor_plan/templates/nautobot_floor_plan/floorplantile_retrieve.html [hunk 19]
```diff
@@ -37,10 +37,18 @@
                     </span>
                 </td>
             </tr>
+            <tr>
+                <td>Tile Type</td>
```
Comment (glennmatthews) [inline nautobot_floor_plan/templates/nautobot_floor_plan/floorplantile_retrieve.html:41]: Maybe `Allocation Type`?

File: nautobot_floor_plan/choices.py [hunk 20]
```diff
@@ -29,3 +29,15 @@ class AxisLabelsChoices(ChoiceSet):
...
+
+class AllocationTypeChoices(ChoiceSet):
+    """Choices for tile allocation type."""
+
+    RACK = "rack"
+    RACKGROUP = "rackgroup"
+
+    CHOICES = (
+        (RACK, "rack"),
+        (RACKGROUP, "rackgroup"),
```
Comment (glennmatthews) [inline nautobot_floor_plan/choices.py:42]:
```suggestion
        (RACK, "Rack"),
        (RACKGROUP, "RackGroup"),
```

**Review by glennmatthews — APPROVED** [review]
